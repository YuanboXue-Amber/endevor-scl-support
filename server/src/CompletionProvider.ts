import { CompletionItem, CompletionItemKind } from "vscode-languageserver";
import { ParserTags } from './oldParser/ParserTags';
import { isNullOrUndefined } from "util";

export function composeCompletionItemsFromKeywords(): CompletionItem[] {
    let completionItems: CompletionItem[] = [];

    Object.values(ParserTags).forEach((value) => {
        // key: the name of the object key
        // index: the ordinal position of the key within the object
        if (isNullOrUndefined(value)) {
            return;
        }
        if (value.indexOf(",") > 0) {
            const values = value.split(", ");
            for (const v of values) {
                completionItems.push({
                    label: v.toUpperCase() + " ",
                    kind: CompletionItemKind.Keyword,
                    documentation: "Endevor SCL keyword"
                });
            }
            return;
        }
        completionItems.push({
            label: value.toUpperCase() + " ",
            kind: CompletionItemKind.Keyword,
            documentation: "Endevor SCL keyword"
        });
    });
    return completionItems;
}