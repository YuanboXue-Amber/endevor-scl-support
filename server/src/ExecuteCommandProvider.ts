import { ExecuteCommandParams, Command, TextDocumentIdentifier } from "vscode-languageserver";
import { SCLDocument, SCLstatement } from './documents/SCLDocument';
import * as fs from 'fs';
import * as path from 'path';
import { exec, execSync } from "child_process";

const SUBMITSCL_COMMAND = Command.create('Submit scl', "endevorscl.submitscl");
const OPENDOC_COMMAND = Command.create('Open link to techDocs', "endevorscl.techdocs");
export const commands = [
    SUBMITSCL_COMMAND,
    OPENDOC_COMMAND
];

export const executeSubmitSCL= ((document: SCLDocument, starti: number) => {
    for (const statement of document.statements) {
        if (statement.starti === starti) {
            const sclString = document.textDocument.getText().substring(starti, statement.endi);
            const sclFilePath = path.join(__dirname, "temp.txt");
            fs.writeFileSync(sclFilePath, sclString, "utf-8");
            const result = execSync(`zowe endevor submit scl --sf ${sclFilePath} --sclt list -i CMEWXYTS`);
            return result;
        }
    }
    throw new Error(`Unable to locate codeLens request in the current document`);
});
