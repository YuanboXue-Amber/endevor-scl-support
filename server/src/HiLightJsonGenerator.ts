import { isNullOrUndefined } from "util";
import { ParserTags } from './parser/parserTags';
import * as fs from 'fs';

const level0keywords: string[] = [
    ParserTags.SET,
    ParserTags.ADD,
    ParserTags.ARCHIVE,
    ParserTags.COPY,
    ParserTags.DELETE,
    ParserTags.GENERATE,
    ParserTags.LIST,
    ParserTags.MOVE,
    ParserTags.PRINT,
    ParserTags.RESTORE,
    ParserTags.RETRIEVE,
    ParserTags.SIGNIN,
    ParserTags.TRANSFER,
    ParserTags.UPDATE,
    ParserTags.VALIDATE
];

const level1keywords: string[] = [
    ParserTags.ACTION,
    ParserTags.BUILD,
    ParserTags.FROM,
    ParserTags.OPTIONS,
    ParserTags.OPTION,
    ParserTags.TO,
    ParserTags.WHERE,
    ParserTags.STOPRC,
    ParserTags.ELEMENT,
    ParserTags.THROUGH,
];

const level2keywords: string[] = [
    ParserTags.FILE,
    ParserTags.DDNAME,
    ParserTags.DSNAME,
    ParserTags.PATH,
    ParserTags.MEMBER,
    ParserTags.USSFILE,

    ParserTags.ENVIRONMENT,
    ParserTags.SYSTEM,
    ParserTags.SUBSYSTEM,
    ParserTags.TYPE,
    ParserTags.STAGE,
    ParserTags.NUMBER,

    ParserTags.CCID,
    ParserTags.COMMENT,
    ParserTags.NEW,
    ParserTags.VERSION,
    ParserTags.IF,
    ParserTags.PRESENT,
    ParserTags.INPUT,
    ParserTags.SOURCE,
    ParserTags.OVERRIDE,
    ParserTags.SIGNOUT,
    ParserTags.BYPASS,
    ParserTags.PROCESSOR,
    ParserTags.GROUP,
    ParserTags.EQUAL,
    ParserTags.AUTOGEN,
    ParserTags.SPAN,
    ParserTags.NONE,
    ParserTags.ALL,
    ParserTags.SYSTEMS,
    ParserTags.SUBSYSTEMS
];

// compose a regex from parsetag. Eg. for APPROVER it is APP(ROVER|ROVE|ROV|RO|R|\b)
const composeRegex = ((tagValue: string): string | undefined => {

    const kwMatchArray = tagValue.match(/([A-Z]*)([a-zA-Z]*)/);
    if (isNullOrUndefined(kwMatchArray)) {
        return undefined;
    }
    const mandatoryText = kwMatchArray[1];
    const optionalText = kwMatchArray[2].toUpperCase();

    let regexString = mandatoryText + "(";
    for (let i = optionalText.length; i > 0; -- i) {
        regexString += optionalText.substr(0, i) + "|";
    }
    regexString += "\\b)";
    return `\\b(${regexString})\\b`;
});

const jsonGenerator = (() => {
    const textMateJson: string = fs.readFileSync("./syntaxes/scl.tmLanguage.json", "UTF-8")
    const textMateObj = JSON.parse(textMateJson);
    textMateObj.repository.keywords.patterns = [];

    level0keywords.forEach((kw) => {
        const regex = composeRegex(kw);
        if (isNullOrUndefined(regex))
            return;
        textMateObj.repository.keywords.patterns.push({
            name: "keyword.control.scl",
            match: regex
        });
    });

    level1keywords.forEach((kw) => {
        const regex = composeRegex(kw);
        if (isNullOrUndefined(regex))
            return;
        textMateObj.repository.keywords.patterns.push({
            name: "support.class.secondarykeyword.scl",
            match: regex
        });
    });

    level2keywords.forEach((kw) => {
        const regex = composeRegex(kw);
        if (isNullOrUndefined(regex))
            return;
        textMateObj.repository.keywords.patterns.push({
            name: "support.variable.tertiarykeyword.scl",
            match: regex
        });
    });
    fs.writeFileSync("./syntaxes/scl.tmLanguage.json", JSON.stringify(textMateObj, null, 4), "UTF-8");
});

jsonGenerator();