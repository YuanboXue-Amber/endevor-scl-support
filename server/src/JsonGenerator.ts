import { isNullOrUndefined } from "util";
import { ParserTags } from './parser/parserTags';
import * as fs from 'fs';

const level0keywords: string[] = [
    ParserTags.T_SET,
    ParserTags.T_ADD,
    ParserTags.T_ARCHIVE,
    ParserTags.T_COPY,
    ParserTags.T_DELETE,
    ParserTags.T_GENERATE,
    ParserTags.T_LIST,
    ParserTags.T_MOVE,
    ParserTags.T_PRINT,
    ParserTags.T_RESTORE,
    ParserTags.T_RETRIEVE,
    ParserTags.T_SIGNIN,
    ParserTags.T_TRANSFER,
    ParserTags.T_UPDATE,
    ParserTags.T_VALIDATE
];

const level1keywords: string[] = [
    ParserTags.T_ACTION,
    ParserTags.T_BUILD,
    ParserTags.T_FROM,
    ParserTags.T_OPTIONS,
    ParserTags.T_TO,
    ParserTags.T_WHERE,
    ParserTags.T_STOPRC,
    ParserTags.T_ELEMENT,
    ParserTags.T_THROUGH,
];

const level2keywords: string[] = [
    ParserTags.T_FILE,
    ParserTags.T_DDNAME,
    ParserTags.T_DSNAME,
    ParserTags.T_PATH,
    ParserTags.T_MEMBER,
    ParserTags.T_USSFILE,

    ParserTags.T_ENVIRONMENT,
    ParserTags.T_SYSTEM,
    ParserTags.T_SUBSYSTEM,
    ParserTags.T_TYPE,
    ParserTags.T_STAGE,
    ParserTags.T_NUMBER,
];

// compose a regex from parsetag. Eg. for T_APPROVER it is APP(ROVER|ROVE|ROV|RO|R|\b)
const composeRegex = ((tagValue: string): string | undefined => {

    const kwMatchArray = tagValue.match(/([A-Z]*)([a-zA-Z]*)/);
    if (isNullOrUndefined(kwMatchArray)) {
        return undefined;
    }
    const mandatoryText = kwMatchArray[1];
    const optionalText = kwMatchArray[2].toUpperCase();

    let regexString = mandatoryText + "(";
    for (let i = optionalText.length; i > 0; -- i) {
        regexString += optionalText.substr(0, i) + "|";
    }
    regexString += "\\b)";
    return `\\b(${regexString})\\b`;
});

const jsonGenerator = (() => {
    const textMateJson: string = fs.readFileSync("./syntaxes/scl.tmLanguage.json", "UTF-8")
    const textMateObj = JSON.parse(textMateJson);
    textMateObj.repository.keywords.patterns = [];

    level0keywords.forEach((kw) => {
        const regex = composeRegex(kw);
        if (isNullOrUndefined(regex))
            return;
        textMateObj.repository.keywords.patterns.push({
            name: "keyword.control.scl",
            match: regex
        });
    });

    level1keywords.forEach((kw) => {
        const regex = composeRegex(kw);
        if (isNullOrUndefined(regex))
            return;
        textMateObj.repository.keywords.patterns.push({
            name: "support.class.secondarykeyword.scl",
            match: regex
        });
    });

    level2keywords.forEach((kw) => {
        const regex = composeRegex(kw);
        if (isNullOrUndefined(regex))
            return;
        textMateObj.repository.keywords.patterns.push({
            name: "support.variable.tertiarykeyword.scl",
            match: regex
        });
    });
    fs.writeFileSync("./syntaxes/scl.tmLanguage.json", JSON.stringify(textMateObj, null, 4), "UTF-8");
});

jsonGenerator();