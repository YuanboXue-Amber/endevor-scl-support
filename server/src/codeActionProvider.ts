import { CodeAction, CodeActionParams, Diagnostic, DiagnosticSeverity, CodeActionKind, TextDocument } from 'vscode-languageserver';
import { isNullOrUndefined } from "util";

export function quickfix(textDocument: TextDocument, parms: CodeActionParams): CodeAction[]{
    const diagnostics = parms.context.diagnostics;
    if (isNullOrUndefined(diagnostics) || diagnostics.length === 0) {
        return [];
    }
    const test: Diagnostic = diagnostics[0];

    const codeActions: CodeAction[] = [];
    if (test.severity === DiagnosticSeverity.Error) { // Warning &&
        // diagnostics[0].message.indexOf("owercase") >= 0) {
        codeActions.push({
            title: "Uppercase the keyword",
            kind: CodeActionKind.QuickFix,
            diagnostics: [test],
            edit: {
                changes: {
                    [parms.textDocument.uri]: [{
                        range: test.range, newText: textDocument.getText(test.range).toUpperCase()
                    }]
                }
            }
        });
    }
    return codeActions;
}
