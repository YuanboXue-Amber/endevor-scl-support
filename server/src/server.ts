import {
    createConnection,
    TextDocuments,
    ProposedFeatures,
    InitializeParams,
    CompletionItem,
    TextDocumentPositionParams,
    InitializeResult,
    CodeActionKind,
    CodeActionParams,
    CodeAction,
    TextDocumentChangeEvent,
    TextDocumentSyncKind,
    DidOpenTextDocumentParams,
    TextDocumentItem,
    DidCloseTextDocumentParams,
    DidChangeTextDocumentParams,
    Diagnostic,
    DidChangeConfigurationNotification,
    TextDocument,
    ExecuteCommandParams,
    CodeLensParams,
    Command,
    TextDocumentIdentifier
} from 'vscode-languageserver';
import { isNullOrUndefined } from 'util';
import { quickfix } from './CodeActionProvider';
import { composeCompletionItemsFromKeywords } from './CompletionProvider';
import { SCLDocumentManager, IDocumentSettings } from './documents/SCLDocumentManager';
import { SCLDocument } from './documents/SCLDocument';
import { prepareTrees } from './parser/syntaxTrees/PrepareTrees';
import { commands } from './ExecuteCommandProvider';

// Create a connection for the server. The connection uses Node's IPC as a transport.
// Also include all preview / proposed LSP features.
let connection = createConnection(ProposedFeatures.all);

let documents: TextDocuments = new TextDocuments();

const documentManager = new SCLDocumentManager();

// Does the clients accepts diagnostics with related information?
let hasConfigurationCapability: boolean = false;
let hasDiagnosticRelatedInformationCapability: boolean = false;
let hasCodeActionLiteralsCapability: boolean = false;

let completionItems: CompletionItem[] = [];

connection.onInitialize((params: InitializeParams) => {
    let capabilities = params.capabilities;

    hasConfigurationCapability = !!(
        capabilities.workspace && !!capabilities.workspace.configuration
    );

    hasDiagnosticRelatedInformationCapability = !!(
        capabilities.textDocument &&
        capabilities.textDocument.publishDiagnostics &&
        capabilities.textDocument.publishDiagnostics.relatedInformation
    );

    hasCodeActionLiteralsCapability = !!(
      capabilities.textDocument &&
      capabilities.textDocument.codeAction &&
      capabilities.textDocument.codeAction.codeActionLiteralSupport
    );

    SCLDocumentManager.capabilities = {
        hasDiagnosticRelatedInformationCapability
    };

    const result: InitializeResult = {
        capabilities: {
            // textDocumentSync: TextDocumentSyncKind.Incremental,
            textDocumentSync: documents.syncKind,
            // Tell the client that the server supports code completion
            completionProvider: {
                resolveProvider: true
            },
            documentFormattingProvider: true,
            executeCommandProvider: {
                commands: commands.map(command => command.command)
            },
            codeLensProvider: {
                resolveProvider: true
            }
            // hoverProvider: true, // AmberTODO
        }
    };

    if (hasCodeActionLiteralsCapability) {
        result.capabilities.codeActionProvider = {
            codeActionKinds: [CodeActionKind.QuickFix]
        };
    }

    return result;
});

// After the client received the result of the initialize request
// but before the client is sending any other request or notification to the server
connection.onInitialized(() => {
    prepareTrees(); // read syntax trees from Json
    completionItems = composeCompletionItemsFromKeywords(); // initialize completion items
    if (hasConfigurationCapability) {
        // Register for all configuration changes.
        connection.client.register(DidChangeConfigurationNotification.type, undefined);
    }
});

// The global settings, used when the `workspace/configuration` request is not supported by the client.
// Please note that this is not the case when using this server with the client provided in this example
// but could happen with other clients.
const defaultSettings: IDocumentSettings = { maxNumberOfProblems: 1000, isREST: false };
let globalSettings: IDocumentSettings = defaultSettings;

// Cache the settings of all open documents
// let documentSettings: Map<string, Thenable<IDocumentSettings>> = new Map();

let documentSettings: Map<string, Thenable<IDocumentSettings>> = new Map();
connection.onDidChangeConfiguration(change => {
    if (hasConfigurationCapability) {
        // Reset all cached document settings
        documentSettings.clear();
        SCLDocumentManager.config = globalSettings;
        SCLDocumentManager.numberOfProblems = 0;
    } else {
        globalSettings = <IDocumentSettings>(
            (change.settings.languageServerExample || defaultSettings)
        );
    }

    // Revalidate all open text documents
    documents.all().forEach(validateTextDocument);
});

// connection.onDidOpenTextDocument((parm: DidOpenTextDocumentParams) => {
//     const document: SCLDocument = documentManager.openOrChangeDocument(parm.textDocument);

//     const diagnostics: Diagnostic[] = [];
//     document.statements.forEach((statement) => {
//         statement.diagnostics.forEach((diag) => {
//             diagnostics.push(diag.diagnostic);
//         });
//     });
//     connection.sendDiagnostics({
//         uri: document.textDocument.uri,
//         diagnostics
//     });
// });

// connection.onDidChangeTextDocument((parm: DidChangeTextDocumentParams) => {
//     const document: SCLDocument = documentManager.updateDocument(parm.textDocument, parm.contentChanges);

//     const diagnostics: Diagnostic[] = [];
//     document.statements.forEach((statement) => {
//         statement.diagnostics.forEach((diag) => {
//             diagnostics.push(diag.diagnostic);
//         });
//     });
//     connection.sendDiagnostics({
//         uri: document.textDocument.uri,
//         diagnostics
//     });
// });

// connection.onDidCloseTextDocument((parm: DidCloseTextDocumentParams) => {
//     documentManager.closeDocument(parm.textDocument.uri);
// });

// Only keep settings for open documents
documents.onDidClose(e => {
    documentSettings.delete(e.document.uri);
    documentManager.closeDocument(e.document.uri);
});

function getDocumentSettings(resource: string): Thenable<IDocumentSettings> {
    if (!hasConfigurationCapability) {
        return Promise.resolve(globalSettings);
    }
    let result = documentSettings.get(resource);
    if (!result) {
        result = connection.workspace.getConfiguration({
            scopeUri: resource,
            section: 'endevorSclLanguageServer'
        });
        documentSettings.set(resource, result);
    }
    return result;
}

// The content of a text document has changed. This event is emitted
// when the text document first opened or when its content has changed.
documents.onDidChangeContent(change => {
    validateTextDocument(change.document);
});

async function validateTextDocument(textDocument: TextDocument): Promise<void> {
    SCLDocumentManager.config = await getDocumentSettings(textDocument.uri);

    const document: SCLDocument = documentManager.openOrChangeDocument(textDocument);

    const diagnostics: Diagnostic[] = [];
    document.statements.forEach((statement) => {
        statement.diagnostics.forEach((diag) => {
            diagnostics.push(diag.diagnostic);
        });
    });
    connection.sendDiagnostics({
        uri: document.textDocument.uri,
        diagnostics
    });
}


connection.onCodeAction(provideCodeActions);

async function provideCodeActions(parms: CodeActionParams): Promise<CodeAction[]> {
    if (!parms.context.diagnostics.length) {
        return [];
    }
    const document = documentManager.documents.get(parms.textDocument.uri);
    if (isNullOrUndefined(document)) {
        return [];
    }
    return quickfix(document.textDocument, parms);
}

// This handler provides the initial list of the completion items.
connection.onCompletion(
    (_textDocumentPosition: TextDocumentPositionParams): CompletionItem[] => {
        const completionItemsBySyntax = documentManager.getCompletionBySyntax(_textDocumentPosition);
        if (completionItemsBySyntax.length > 0)
            return completionItemsBySyntax;
        return []; // completionItems;
    }
);

// A void completion resolve function just to handle the completion/resolve request.
connection.onCompletionResolve(
    (item: CompletionItem): CompletionItem => {
        return item;
    }
);

connection.onDocumentFormatting(evt => documentManager.formatDocument(evt.textDocument));

connection.onCodeLens(evt => documentManager.computeCodeLenses(evt.textDocument));

connection.onCodeLensResolve(codeLens => codeLens);
connection.onExecuteCommand(async (params: ExecuteCommandParams) => {
    connection.window.showInformationMessage('Hello World!');
    const result = await documentManager.executeCodeLens(params);
    connection.console.log("result is: " + result);
    connection.window.showInformationMessage("result is: " + result);
    connection.window.showErrorMessage("TEST AMBER ERROR");
    connection.window.showWarningMessage("TEST AMBER WARN");
});

documents.listen(connection);

// Listen on the connection
connection.listen();

// connection.window.showInformationMessage('Hello hello World!');