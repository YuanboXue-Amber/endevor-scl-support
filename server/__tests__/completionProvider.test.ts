import { composeCompletionItemsFromKeywords } from '../src/completionProvider';

describe("Test", () => {
    it("", async () => {
        expect(composeCompletionItemsFromKeywords()).toMatchSnapshot();
    });
});