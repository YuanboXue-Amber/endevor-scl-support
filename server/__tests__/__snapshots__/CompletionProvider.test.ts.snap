// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test keyword completion return a list of completion item from keywords in ParserTags 1`] = `
Array [
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "ACTION",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "ADD",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "ARCHIVE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "BUILD",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "COPY",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "DDNAME",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "DELETE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "DSNAME",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "ELEMENT",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "ENVIRONMENT",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "FILE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "FROM",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "GENERATE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "LIST",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "MEMBER",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "MOVE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "NUMBER",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "OPTIONS",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "OPTION",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "PATH",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "PRINT",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "RESTORE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "RETRIEVE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "SET",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "SIGNIN",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "STAGE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "STOPRC",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "SUBSYSTEM",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "SYSTEM",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "THROUGH",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "TRANSFER",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "TO",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "TYPE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "UPDATE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "USSFILE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "VALIDATE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "WHERE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "CCID",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "COMMENT",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "NEW",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "VERSION",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "IF",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "PRESENT",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "INPUT",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "SOURCE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "OVERRIDE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "SIGNOUT",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "BYPASS",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "PROCESSOR",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "GROUP",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "EQUAL",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "AUTOGEN",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "SPAN",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "NONE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "ALL",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "SYSTEMS",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "SUBSYSTEMS",
  },
]
`;
